File: scripts/cloud/managed-sandbox-guest-interrupt.mjs
#!/usr/bin/env node

import { existsSync, readFileSync, readdirSync } from "node:fs";
import { pathToFileURL } from "node:url";

const TURN_ROOT = "/var/lib/openagents/managed-sandbox-turns";
const WAIT_MILLIS = 5_000;
const POLL_MILLIS = 50;

const fail = (reason) => {
  process.stderr.write(
    `${JSON.stringify({ error: "managed_sandbox_interrupt_failed", reason })}\n`,
  );
  process.exit(2);
};

const sleep = (millis) => new Promise((resolve) => setTimeout(resolve, millis));

// Read every live process as (pid, processGroupId, sessionId). Zombies are
// excluded: they hold no resources and cannot be signalled.
const liveProcesses = () => {
  if (process.platform !== "linux" || !existsSync("/proc")) fail("linux_proc_required");
  const observed = [];
  for (const entry of readdirSync("/proc", { withFileTypes: true })) {
    if (!entry.isDirectory() || !/^\d+$/u.test(entry.name)) continue;
    try {
      const stat = readFileSync(`/proc/${entry.name}/stat`, "utf8");
      const close = stat.lastIndexOf(")");
      if (close < 0) fail("process_stat_invalid");
      const fields = stat.slice(close + 2).split(" ");
      const state = fields[0];
      if (state === "Z") continue;
      const processGroupId = Number.parseInt(fields[2] ?? "", 10);
      const sessionId = Number.parseInt(fields[3] ?? "", 10);
      if (!Number.isSafeInteger(processGroupId) || !Number.isSafeInteger(sessionId)) {
        fail("process_stat_invalid");
      }
      observed.push({ pid: Number.parseInt(entry.name, 10), processGroupId, sessionId });
    } catch (error) {
      if (error?.code !== "ENOENT") fail("process_observation_failed");
    }
  }
  return observed;
};

const processGroupMembers = (processGroupId) =>
  liveProcesses()
    .filter((observed) => observed.processGroupId === processGroupId)
    .map((observed) => observed.pid);

// A descendant that calls setpgid() leaves the process group and therefore
// survives `kill(-pgid)` unseen. It stays in the session, so the session is
// what exposes it. Sessions observed before the signal are the ones that
// belong to this turn.
// Both selectors are pure over an observed process table so they can be
// falsified on any platform; production passes the live /proc reading.
export const sessionsIn = (processes, processGroupId) => [
  ...new Set(
    processes
      .filter((observed) => observed.processGroupId === processGroupId)
      .map((observed) => observed.sessionId),
  ),
];

export const escapedDescendantsIn = (processes, processGroupId, sessions, selfPid) =>
  processes.filter(
    (observed) =>
      observed.processGroupId !== processGroupId &&
      sessions.includes(observed.sessionId) &&
      observed.pid !== selfPid &&
      observed.sessionId !== 0,
  );

const sessionsOf = (processGroupId) => sessionsIn(liveProcesses(), processGroupId);

const escapedDescendants = (processGroupId, sessions) =>
  escapedDescendantsIn(liveProcesses(), processGroupId, sessions, process.pid);

const waitForExit = async (processGroupId) => {
  const deadline = Date.now() + WAIT_MILLIS;
  while (Date.now() < deadline) {
    if (processGroupMembers(processGroupId).length === 0) return true;
    await sleep(POLL_MILLIS);
  }
  return processGroupMembers(processGroupId).length === 0;
};

const signal = (processGroupId, signalName) => {
  try {
    process.kill(-processGroupId, signalName);
  } catch (error) {
    if (error?.code !== "ESRCH") fail("process_group_signal_failed");
  }
};

export const interruptProcessGroup = async (processGroupId) => {
  if (!Number.isSafeInteger(processGroupId) || processGroupId < 2) {
    fail("process_group_identity_invalid");
  }
  let escalatedToSigkill = false;
  // Capture the sessions while the group is still intact, so a descendant that
  // leaves the group afterwards is still attributable to this turn.
  const sessions = sessionsOf(processGroupId);
  if (processGroupMembers(processGroupId).length > 0) {
    signal(processGroupId, "SIGTERM");
    if (!(await waitForExit(processGroupId))) {
      escalatedToSigkill = true;
      signal(processGroupId, "SIGKILL");
      if (!(await waitForExit(processGroupId))) fail("process_group_still_active");
    }
  }

  // Signal anything that escaped the group but remains in an observed session,
  // then measure. Every field below is the final observation, not an assertion
  // that the wait loop above must have succeeded.
  for (const escapee of escapedDescendants(processGroupId, sessions)) {
    try {
      process.kill(escapee.pid, "SIGKILL");
    } catch (error) {
      if (error?.code !== "ESRCH") fail("escaped_descendant_signal_failed");
    }
  }
  const deadline = Date.now() + WAIT_MILLIS;
  let remaining = escapedDescendants(processGroupId, sessions);
  while (remaining.length > 0 && Date.now() < deadline) {
    await sleep(POLL_MILLIS);
    remaining = escapedDescendants(processGroupId, sessions);
  }
  const groupRemaining = processGroupMembers(processGroupId).length;
  const descendantsRemaining = remaining.length;
  if (groupRemaining !== 0) fail("process_group_still_active");
  if (descendantsRemaining !== 0) fail("escaped_descendants_still_active");

  // The emitted shape is unchanged: consumers pin `zeroProcessGroup: true` and
  // `descendantsRemaining: 0`. What changed is that both are now the measured
  // result of the observations above rather than literals, and the refusals
  // above guarantee the pinned values can only be reached honestly.
  return {
    schemaVersion: "openagents.managed_sandbox_interrupt_proof.v1",
    processGroupId,
    zeroProcessGroup: groupRemaining === 0,
    descendantsRemaining,
    escalatedToSigkill,
  };
};

if (process.argv[1] !== undefined && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const [turnDirectory] = process.argv.slice(2);
  if (
    typeof turnDirectory !== "string" ||
    !new RegExp(`^${TURN_ROOT}/[a-f0-9]{24}$`, "u").test(turnDirectory)
  ) {
    fail("turn_directory_invalid");
  }
  let processGroupId;
  try {
    processGroupId = Number.parseInt(readFileSync(`${turnDirectory}/pgid`, "utf8"), 10);
  } catch {
    fail("process_group_identity_missing");
  }
  process.stdout.write(`${JSON.stringify(await interruptProcessGroup(processGroupId))}\n`);
}

File: scripts/cloud/managed-sandbox-guest-interrupt.py
#!/usr/bin/env python3
"""Interrupt a managed-sandbox guest turn process group and emit a proof."""

from __future__ import annotations

import json
import os
import re
import signal
import sys
import time
from pathlib import Path
from typing import Any


TURN_ROOT = "/var/lib/openagents/managed-sandbox-turns"
WAIT_SECONDS = 5.0
POLL_SECONDS = 0.05


def fail(reason: str) -> None:
    sys.stderr.write(json.dumps({"error": "managed_sandbox_interrupt_failed", "reason": reason}) + "\n")
    raise SystemExit(2)


def live_processes() -> list[dict[str, int]]:
    if sys.platform != "linux" or not Path("/proc").exists():
        fail("linux_proc_required")
    observed: list[dict[str, int]] = []
    for entry in Path("/proc").iterdir():
        if not entry.is_dir() or not entry.name.isdigit():
            continue
        try:
            stat = (entry / "stat").read_text(encoding="utf-8")
        except OSError as error:
            if getattr(error, "errno", None) != 2:
                fail("process_observation_failed")
            continue
        close = stat.rfind(")")
        if close < 0:
            fail("process_stat_invalid")
        fields = stat[close + 2 :].split(" ")
        if fields[0] == "Z":
            continue
        try:
            process_group_id = int(fields[2])
            session_id = int(fields[3])
        except (IndexError, ValueError):
            fail("process_stat_invalid")
        observed.append(
            {
                "pid": int(entry.name),
                "processGroupId": process_group_id,
                "sessionId": session_id,
            }
        )
    return observed


def sessions_in(processes: list[dict[str, int]], process_group_id: int) -> list[int]:
    seen: list[int] = []
    for observed in processes:
        if observed["processGroupId"] == process_group_id and observed["sessionId"] not in seen:
            seen.append(observed["sessionId"])
    return seen


def escaped_descendants_in(
    processes: list[dict[str, int]],
    process_group_id: int,
    sessions: list[int],
    self_pid: int,
) -> list[dict[str, int]]:
    return [
        observed
        for observed in processes
        if observed["processGroupId"] != process_group_id
        and observed["sessionId"] in sessions
        and observed["pid"] != self_pid
        and observed["sessionId"] != 0
    ]


def process_group_members(process_group_id: int) -> list[int]:
    return [
        observed["pid"]
        for observed in live_processes()
        if observed["processGroupId"] == process_group_id
    ]


def wait_for_exit(process_group_id: int) -> bool:
    deadline = time.monotonic() + WAIT_SECONDS
    while time.monotonic() < deadline:
        if not process_group_members(process_group_id):
            return True
        time.sleep(POLL_SECONDS)
    return not process_group_members(process_group_id)


def signal_group(process_group_id: int, signum: int) -> None:
    try:
        os.killpg(process_group_id, signum)
    except ProcessLookupError:
        pass
    except OSError:
        fail("process_group_signal_failed")


def interrupt_process_group(process_group_id: int) -> dict[str, Any]:
    if not isinstance(process_group_id, int) or process_group_id < 2:
        fail("process_group_identity_invalid")
    escalated_to_sigkill = False
    sessions = sessions_in(live_processes(), process_group_id)
    if process_group_members(process_group_id):
        signal_group(process_group_id, signal.SIGTERM)
        if not wait_for_exit(process_group_id):
            escalated_to_sigkill = True
            signal_group(process_group_id, signal.SIGKILL)
            if not wait_for_exit(process_group_id):
                fail("process_group_still_active")
    for escapee in escaped_descendants_in(
        live_processes(), process_group_id, sessions, os.getpid()
    ):
        try:
            os.kill(escapee["pid"], signal.SIGKILL)
        except ProcessLookupError:
            pass
        except OSError:
            fail("escaped_descendant_signal_failed")
    deadline = time.monotonic() + WAIT_SECONDS
    remaining = escaped_descendants_in(
        live_processes(), process_group_id, sessions, os.getpid()
    )
    while remaining and time.monotonic() < deadline:
        time.sleep(POLL_SECONDS)
        remaining = escaped_descendants_in(
            live_processes(), process_group_id, sessions, os.getpid()
        )
    group_remaining = len(process_group_members(process_group_id))
    descendants_remaining = len(remaining)
    if group_remaining != 0:
        fail("process_group_still_active")
    if descendants_remaining != 0:
        fail("escaped_descendants_still_active")
    return {
        "schemaVersion": "openagents.managed_sandbox_interrupt_proof.v1",
        "processGroupId": process_group_id,
        "zeroProcessGroup": group_remaining == 0,
        "descendantsRemaining": descendants_remaining,
        "escalatedToSigkill": escalated_to_sigkill,
    }


def main() -> None:
    if len(sys.argv) != 2:
        fail("turn_directory_invalid")
    turn_directory = sys.argv[1]
    if not re.fullmatch(rf"{TURN_ROOT}/[a-f0-9]{{24}}", turn_directory):
        fail("turn_directory_invalid")
    try:
        process_group_id = int(Path(turn_directory, "pgid").read_text(encoding="utf-8"))
    except (OSError, ValueError):
        fail("process_group_identity_missing")
    sys.stdout.write(json.dumps(interrupt_process_group(process_group_id)) + "\n")


if __name__ == "__main__":
    main()
